iree/compiler/Dialect/Flow/Transforms/UnrollReductions.cpp

```diff
@@ -0,0 +1,180 @@
+  auto windowDilations = reduceOp.window_dilationsAttr();
+  SmallVector<WindowTuple, 4> sortedWindowAttrs;
+  for (uint32_t i = 0; i < windowDimensions.getNumElements(); ++i) {
+    int64_t windowDimension = windowDimensions.getValue<int64_t>({i});
+    int64_t windowStride = windowStrides.getValue<int64_t>({i});
+    int64_t baseDilation = baseDilations.getValue<int64_t>({i});
+    int64_t windowDilation = windowDilations.getValue<int64_t>({i});
+    sortedWindowAttrs.push_back(WindowTuple(windowDimension, windowStride,
+                                            baseDilation, windowDilation));
+  }
+  llvm::sort(sortedWindowAttrs, [](WindowTuple a, WindowTuple b) {
+    return (std::get<0>(a) - std::get<0>(b)) > 0;
+  });
+
+  // Insert at the same place as the original op.
+  OpBuilder builder(reduceOp);
+  Value temp = reduceOp.operand();
+  for (auto windowAttrs : llvm::enumerate(sortedWindowAttrs)) {
+    int64_t windowDimension = std::get<0>(windowAttrs.value());
+    int64_t windowStride = std::get<1>(windowAttrs.value());
+    int64_t baseDilation = std::get<2>(windowAttrs.value());
+    int64_t windowDilation = std::get<3>(windowAttrs.value());
+
+    // Compute the new result.
+    // NOTE: this should not be required and instead be on the HLO op.
+    auto tempType = temp.getType().cast<RankedTensorType>();
+    SmallVector<int64_t, 4> newShape;
+    for (int64_t i = 0; i < tempType.getRank(); ++i) {
+      if (i != windowDimension) {
+        newShape.push_back(tempType.getDimSize(i));
+      }
+    }
+    auto resultType =
+        RankedTensorType::get(newShape, tempType.getElementType());
+
+    // Create the new reduction using the results of the previous operation.
+    auto singleAttrType =
+        RankedTensorType::get({1}, builder.getIntegerType(64));
+    auto singleReduceOp = builder.create<xla_hlo::ReduceWindowOp>(
+        reduceOp.getLoc(), resultType, temp, reduceOp.init_value(),
+        DenseIntElementsAttr::get(
+            singleAttrType,
+            {builder.getI64IntegerAttr(windowDimension).cast<Attribute>()}),
+        DenseIntElementsAttr::get(
+            singleAttrType,
+            {builder.getI64IntegerAttr(windowStride).cast<Attribute>()}),
+        DenseIntElementsAttr::get(
+            singleAttrType,
+            {builder.getI64IntegerAttr(baseDilation).cast<Attribute>()}),
+        DenseIntElementsAttr::get(
+            singleAttrType,
+            {builder.getI64IntegerAttr(windowDilation).cast<Attribute>()}),
+        reduceOp.paddingAttr());
+    BlockAndValueMapping mapping;
+    reduceOp.body().cloneInto(&singleReduceOp.body(), mapping);
+    temp = singleReduceOp.getResult();
+  }
+
+  // Replace uses of the existing results with the new results.
+  reduceOp.getResult().replaceAllUsesWith(temp);
+
+  // Erase original op.
+  reduceOp.erase();
+}
+
+}  // namespace
+
+class UnrollReductionsPass : public FunctionPass<UnrollReductionsPass> {
```

Comment:
Why is this a custom pass instead of using patterns? The more we do custom stuff the less we can benefit from shared infra, so I'd bias heavily towards using the OpRewritePattern and such

Reply:
¯\_(ツ)_/¯ can be a conversion in the future, but I'm trying to avoid large non-trivial conversions due to all of the issues with debugging the conversion framework. this is nice to work with, very clear as to what and when it is performing its work, and can be cleanly toggled/tested/etc.

Reply:
I don't mean the conversion framework. Just a pattern for an op tranformation with applyPatternsGreedily